sql/2020/07_SEO/alt_images_by_device.sql

```diff
@@ -0,0 +1,78 @@
+#standardSQL
```

Comment:
I have a few suggestions that I'll propose by way of a rewritten query.

- I agree with your comment about using percentiles. For this query it makes sense to represent the % of images with alt attributes per page as a distribution.
- Do the division in the JS function while you're at it, which also simplifies the return type. Function renamed accordingly.
- You could simplify some of the defensive `if` statements by returning `null` in the `catch` block. Null values are ignored in the percentile aggregation by default.
- Various whitespace tidying.
- Uppercase for all keywords, like `AS`.
- Remove all "TEST" logic.

```sql
#standardSQL
# Distribution of the coverage of alt attributes over all images per page.
CREATE TEMPORARY FUNCTION get_percent_alt_images(markup_string STRING)
RETURNS FLOAT64 LANGUAGE js AS '''
try {
  var markup = JSON.parse(markup_string);
  return markup.images.img.alt.present / markup.images.img.total;
} catch (e) {
  return null;
}
''';

SELECT
  client,
  percentile,
  ROUND(APPROX_QUANTILES(pct_alt, 1000)[OFFSET(percentile * 10)], 4) AS median_pct_alt
FROM ( 
  SELECT 
    _TABLE_SUFFIX AS client,
    get_percent_alt_images(JSON_EXTRACT_SCALAR(payload, '$._markup')) AS pct_alt
  FROM
    `httparchive.pages.2020_08_01_*`),
  UNNEST([10, 25, 50, 75, 90]) AS percentile
GROUP BY
  client,
  percentile
ORDER BY
  percentile,
  client
```

Results:


client | percentile | median_pct_alt
-- | -- | --
desktop | 10 | 0.00%
mobile | 10 | 0.00%
desktop | 25 | 10.53%
mobile | 25 | 9.09%
desktop | 50 | 50.00%
mobile | 50 | 50.00%
desktop | 75 | 89.74%
mobile | 75 | 89.47%
desktop | 90 | 100.00%
mobile | 90 | 100.00%

_(to be honest, I'm very surprised that the median is exactly 50% - please let me know if you spot anything wrong with my query!)_

Reply:
I should have made the first query the last one. It was one I did not think was too valuable, so did not focus on it much.

I think I'd already added it into my percentile query. Will check and fold in your comments there.

> Do the division in the JS function while you're at it, which also simplifies the return type. Function renamed accordingly.

A few reasons I leave some maths to the SQL:

- SQL has a safe div so no exception on zeros. Not an issue with one return value, bot could break all JS result data with a single mistake. 
- Returning more data could come in handy for the SQL writer. e.g. I often return a count and not a boolean indicating non zero. Again, to minify churn if the SQL writer decided they wanted something else. Or to try and keep the data provided consistent between queries.
- I have the feeling that SQL runs far faster then JS.

> You could simplify some of the defensive if statements by returning null in the catch block. Null values are ignored in the percentile aggregation by default.

That script does not go as defensive as a normally go!

This goes against my own beliefs and what I think is best practice. Exceptions should not be used to handle normal code flow.

It's more apparent when you see some of my more complex JS. If I was not defensive, then an exception would mean all subsequent data calculations would be lost. A bug in an early bit of code could cause the loss of much data.

The pattern I use where I create the result object at the start and always return it (even in an exception) is deliberate. That way an exception does not stop all data from being returned (null).

I've also tried to make my exception/error handling provide visible clues in the results that things were broken. Returning something like a null means an issue is silently ignored. Returning an error message where possible really stands out.

That reminds me, I built in exception handling to my custom metrics. I should check that to see if they did have issues. I did that after noticing that a reasonable percentage of results from last year had blank custom metrics, indicating an exception happened and was not handled.

Other comments are sound and I will work on them today.

I'll double check the custom metrics for this to make sure there is nothing strange causing the 50% median.

Reply:
It's hard to get my head around using percents to feed into percentiles 🤯

I've updated pages_markup_by_device_and_percentile.sql to include the percentile version of that query and have it pulling out a few values to help me check validity and understand things. Here's the result from a live run:

Row | percentile | client | total | img_count | images_with_alt_present_percent | images_with_alt_blank_percent | images_with_alt_missing_percent | images_with_alt_present | images_with_alt_blank | images_with_alt_missing |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 10 | desktop | 5593642 | 2 | 0.0 | 0.0 | 0.0 | 0 | 0 | 0 |  
2 | 10 | mobile | 6347640 | 2 | 0.0 | 0.0 | 0.0 | 0 | 0 | 0 |  
3 | 25 | desktop | 5593642 | 8 | 0.1429 | 0.0 | 0.0 | 1 | 0 | 0 |  
4 | 25 | mobile | 6347640 | 8 | 0.125 | 0.0 | 0.0 | 1 | 0 | 0 |  
5 | 50 | desktop | 5593642 | 21 | 0.5385 | 0.1 | 0.0299 | 7 | 2 | 1 |  
6 | 50 | mobile | 6347640 | 19 | 0.5122 | 0.1111 | 0.0244 | 6 | 2 | 1 |  
7 | 75 | desktop | 5593642 | 43 | 0.9091 | 0.4906 | 0.3333 | 22 | 10 | 5 |  
8 | 75 | mobile | 6347640 | 39 | 0.9063 | 0.5 | 0.3333 | 20 | 9 | 5 |  
9 | 90 | desktop | 5593642 | 82 | 1.0 | 0.8261 | 0.8037 | 48 | 24 | 19 |  
10 | 90 | mobile | 6347640 | 73 | 1.0 | 0.8333 | 0.8235 | 43 | 22 | 18

The first column and  last three columns are just counts. The others are the percents.

So I'm guessing it's saying things like (mobile):

The median page has 19 images, 10% have 2 or less, and 10% have 73 or more. 

10% of pages have alt attributes present on all their images? While 10% have no images with the alt present. The median is to have 51% of alts present?

10% have 82% of their images without any alt attribute. 50% do well with less than 3% missing.

50% use less than 10% blank attributes.

Time to rest my brain and clean out all that test code.